File: ts/src/memory.ts
import { createHash } from "node:crypto"
import { existsSync, readFileSync, writeFileSync } from "node:fs"
import { resolve } from "node:path"

const MEMORY_PATH = "docs/memory.md"

export interface ChangeRecord {
  type: string
  original: string
  replacement: string
}

export interface MemoryEntry {
  date: string
  prompt: string
  hash: string
  score: number
  band: string
  action: string
  changes?: ChangeRecord[]
  outcome: string
  note?: string
}

function findMemory(): string | null {
  let dir = process.cwd()
  for (;;) {
    const candidate = resolve(dir, MEMORY_PATH)
    if (existsSync(candidate)) return candidate
    const parent = resolve(dir, "..")
    if (parent === dir) return null
    dir = parent
  }
}

function promptHash(prompt: string): string {
  return createHash("md5").update(prompt, "utf-8").digest("hex").slice(0, 8)
}

function formatEntry(
  prompt: string,
  hash: string,
  score: number,
  band: string,
  action: string,
  outcome: string,
  note: string,
  changes?: ChangeRecord[],
): string {
  const date = new Date().toISOString().replace(/\.\d{3}Z$/, "Z")

  let changesStr = ""
  if (changes && changes.length > 0) {
    const items = changes.map((c) => {
      const orig = c.original.replace(/"/g, "'")
      const repl = c.replacement.replace(/\n/g, " ").replace(/"/g, "'").slice(0, 60)
      return `      - type: ${c.type}\n        original: "${orig}"\n        replacement: "${repl}"`
    })
    changesStr = "    changes:\n" + items.join("\n") + "\n"
  }

  const noteStr = note ? `    note: "${note}"\n` : ""

  return (
    `- date: ${date}\n` +
    `  prompt: "${prompt.slice(0, 80)}"\n` +
    `  hash: ${hash}\n` +
    `  score: ${score.toFixed(1)}\n` +
    `  band: ${band}\n` +
    `  action: ${action}\n` +
    changesStr +
    `  outcome: ${outcome}\n` +
    noteStr
  )
}

export function writeEntry(
  prompt: string,
  action = "none",
  changes?: ChangeRecord[],
  outcome = "accepted",
  note = "",
  score = 0,
  band = "",
): string | null {
  const memPath = findMemory()
  if (!memPath) return null

  const hash = promptHash(prompt)
  const entry = formatEntry(prompt, hash, score, band, action, outcome, note, changes)

  let existing = ""
  try {
    existing = readFileSync(memPath, "utf-8")
  } catch {
    // file doesn't exist yet
  }

  if (!existing.includes("## Entries")) {
    existing += "\n## Entries\n\n"
  }

  const marker = "## Entries"
  const idx = existing.lastIndexOf(marker)
  if (idx === -1) {
    existing += entry + "\n"
  } else {
    existing = existing.slice(0, idx) + marker + "\n\n" + entry + existing.slice(idx + marker.length)
  }

  writeFileSync(memPath, existing, "utf-8")
  return memPath
}

export function readEntries(count = 5): MemoryEntry[] {
  const memPath = findMemory()
  if (!memPath) return []

  let text: string
  try {
    text = readFileSync(memPath, "utf-8")
  } catch {
    return []
  }

  const marker = "## Entries"
  let idx = text.indexOf(marker)
  if (idx === -1) return []
  text = text.slice(idx + marker.length)

  const entries: MemoryEntry[] = []
  let current: Partial<MemoryEntry> = {}

  for (const raw of text.split("\n")) {
    const line = raw.trim()
    if (line.startsWith("- date:")) {
      if (current.date) entries.push(current as MemoryEntry)
      current = { date: line.split(":").slice(1).join(":").trim(), score: 0, band: "" }
    } else if (line.startsWith("prompt:") && current.date) {
      current.prompt = line.split(":").slice(1).join(":").trim().replace(/^"/, "").replace(/"$/, "")
    } else if (line.startsWith("hash:") && current.date) {
      current.hash = line.split(":").slice(1).join(":").trim()
    } else if (line.startsWith("score:") && current.date) {
      current.score = parseFloat(line.split(":").slice(1).join(":").trim()) || 0
    } else if (line.startsWith("band:") && current.date) {
      current.band = line.split(":").slice(1).join(":").trim()
    } else if (line.startsWith("action:") && current.date) {
      current.action = line.split(":").slice(1).join(":").trim()
    } else if (line.startsWith("outcome:") && current.date) {
      current.outcome = line.split(":").slice(1).join(":").trim()
    } else if (line.startsWith("note:") && current.date) {
      current.note = line.split(":").slice(1).join(":").trim().replace(/^"/, "").replace(/"$/, "")
    }
  }
  if (current.date) entries.push(current as MemoryEntry)

  return entries.slice(-count)
}

export function summary(): string {
  const entries = readEntries(20)
  if (entries.length === 0) return "No interactions logged yet."

  const lines = [`Last ${entries.length} interaction(s):`]
  for (const e of entries) {
    const score = e.score
    const action = e.action ?? "?"
    const prompt = (e.prompt ?? "?").slice(0, 50)
    lines.push(`  [${score}/10] ${action}: ${prompt}`)
  }
  return lines.join("\n")
}

File: src/ambiguity/memory.py
"""Project-level memory — logs prompt interactions to docs/memory.md."""

import hashlib
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .analyzer import Analysis


MEMORY_PATH = Path("docs") / "memory.md"


def _find_memory() -> Path | None:
    cwd = Path.cwd()
    for parent in [cwd] + list(cwd.parents):
        candidate = parent / MEMORY_PATH
        if candidate.exists():
            return candidate
    return None


def _prompt_hash(prompt: str) -> str:
    return hashlib.sha256(prompt.encode("utf-8")).hexdigest()[:8]


def _format_entry(
    analysis: Analysis,
    action: str = "none",
    changes: list[dict[str, str]] | None = None,
    outcome: str = "accepted",
    note: str = "",
) -> str:
    ts = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    h = _prompt_hash(analysis.raw_text)
    changes_str = ""
    if changes:
        items = []
        for c in changes:
            orig = c.get("original", "").replace('"', "'")
            repl = c.get("replacement", "").replace("\n", " ").replace('"', "'")[:60]
            items.append(f"      - type: {c['type']}\n        original: \"{orig}\"\n        replacement: \"{repl}\"")
        changes_str = "    changes:\n" + "\n".join(items) + "\n"
    note_str = f"    note: \"{note}\"\n" if note else ""

    return (
        f"- date: {ts}\n"
        f"  prompt: \"{analysis.raw_text[:80]}\"\n"
        f"  hash: {h}\n"
        f"  score: {analysis.score.total:.1f}\n"
        f"  band: {analysis.score.band}\n"
        f"  action: {action}\n"
        f"{changes_str}"
        f"  outcome: {outcome}\n"
        f"{note_str}"
    )


def log_interaction(
    prompt: str,
    action: str = "none",
    changes: list[dict[str, str]] | None = None,
    outcome: str = "accepted",
    note: str = "",
) -> Path | None:
    mem_path = _find_memory()
    if not mem_path:
        return None

    analysis = Analysis(prompt)
    entry = _format_entry(analysis, action, changes, outcome, note)

    try:
        existing = mem_path.read_text(encoding="utf-8")
    except (FileNotFoundError, IOError):
        existing = ""

    if "## Entries" not in existing:
        existing += "\n## Entries\n\n"

    section_marker = "## Entries"
    idx = existing.rfind(section_marker)
    if idx == -1:
        existing += entry + "\n"
    else:
        existing = existing[:idx] + section_marker + "\n\n" + entry + existing[idx + len(section_marker):]

    mem_path.write_text(existing, encoding="utf-8")
    return mem_path


def recent_entries(count: int = 5) -> list[dict[str, Any]]:
    mem_path = _find_memory()
    if not mem_path:
        return []

    try:
        text = mem_path.read_text(encoding="utf-8")
    except (FileNotFoundError, IOError):
        return []

    entries_marker = "## Entries"
    idx = text.find(entries_marker)
    if idx == -1:
        return []
    text = text[idx + len(entries_marker):]

    entries: list[dict[str, Any]] = []
    current: dict[str, Any] = {}
    for line in text.split("\n"):
        line = line.strip()
        if line.startswith("- date:"):
            if current:
                entries.append(current)
            current = {"date": line.split(":", 1)[1].strip()}
        elif line.startswith("prompt:") and current:
            current["prompt"] = line.split(":", 1)[1].strip().strip('"')
        elif line.startswith("score:") and current:
            try:
                current["score"] = float(line.split(":", 1)[1].strip())
            except ValueError:
                pass
        elif line.startswith("action:") and current:
            current["action"] = line.split(":", 1)[1].strip()

    if current:
        entries.append(current)

    return entries[-count:]


def summary() -> str:
    entries = recent_entries(20)
    if not entries:
        return "No interactions logged yet."

    lines = [f"Last {len(entries)} interaction(s):"]
    for e in entries:
        score = e.get("score", "?")
        action = e.get("action", "?")
        prompt = e.get("prompt", "?")[:50]
        lines.append(f"  [{score}/10] {action}: {prompt}")
    return "\n".join(lines)
